src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers/TypeUpgradeAnalyzer.cs

```diff
@@ -0,0 +1,174 @@
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public class TypeUpgradeAnalyzer : DiagnosticAnalyzer
+    {
+        /// <summary>
+        /// The diagnostic ID for diagnostics produced by this analyzer.
+        /// </summary>
+        public const string DiagnosticId = "UA0002";
+
+        /// <summary>
+        /// Key name for the diagnostic property containing the full name of the type
+        /// the code fix provider should use to replace the syntax node identified
+        /// in the diagnostic.
+        /// </summary>
+        public const string NewIdentifierKey = "NewIdentifier";
+
+        /// <summary>
+        /// The diagnsotic category for diagnostics produced by this analyzer.
+        /// </summary>
+        private const string Category = "Upgrade";
+
+        /// <summary>
+        /// The suffix that type map files are expected to use.
+        /// </summary>
+        private const string TypeMapFileSuffix = ".typemap";
+
+        /// <summary>
+        /// An enumerable of <see cref="TypeMapping"/> objects defining
+        /// the type names that should be replaced.
+        /// </summary>
+        private IEnumerable<TypeMapping> _mappings = Enumerable.Empty<TypeMapping>();
+
+        private static readonly LocalizableString Title = new LocalizableResourceString(nameof(Resources.TypeUpgradeTitle), Resources.ResourceManager, typeof(Resources));
+        private static readonly LocalizableString MessageFormat = new LocalizableResourceString(nameof(Resources.TypeUpgradeMessageFormat), Resources.ResourceManager, typeof(Resources));
+        private static readonly LocalizableString Description = new LocalizableResourceString(nameof(Resources.TypeUpgradeDescription), Resources.ResourceManager, typeof(Resources));
+
+        private static readonly DiagnosticDescriptor Rule = new(DiagnosticId, Title, MessageFormat, Category, DiagnosticSeverity.Warning, isEnabledByDefault: true, description: Description);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        /// <summary>
+        /// Initializes the analyzer by registering analysis callback methods.
+        /// </summary>
+        /// <param name="context">The context to use for initialization.</param>
+        public override void Initialize(AnalysisContext context)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                // Load analyzer configuration defining the types that should be mapped.
+                _mappings = LoadMappings(context.Options.AdditionalFiles);
```

Comment:
You've got a race conditiion here. The mappings should be passed through the actions that are registered in case a mapping is changed at some point and then they'd be swapped from underneath you at the next compilation

Reply:
Good point. I don't expect the mappings to change, but it makes sense to pass them along since we're checking here that they're non-empty. I'll make that change.

Reply:
I like the swap to pass-thru for _mappings but would like to ask more about the race condition.

@twsouthwick - are you indicating this analyzer is a singleton? Is there a scenario where an analyzer should be stateful or is that an edge case?